himbaechel/uarch/gowin/gowin.cc

```diff
@@ -319,51 +358,57 @@ void GowinImpl::place_constrained_hclk_cells()
         if ((seen_hclk_users.find(ci->name) != seen_hclk_users.end()))
             continue;
 
-        if (((is_iologici(ci) || is_iologico(ci)) &&
-             !ci->type.in(id_ODDR, id_ODDRC, id_IDDR, id_IDDRC, id_IOLOGICI_EMPTY, id_IOLOGICO_EMPTY))) {
-            NetInfo *hclk_net = ci->getPort(id_FCLK);
-            if (hclk_net)
-                continue;
-            CellInfo *hclk_driver = hclk_net->driver.cell;
-            if (!hclk_driver)
-                continue;
-            if (chip.str(ctx) == "GW1N-9C" && hclk_driver->type != id_CLKDIV2) {
-                // CLKDIV doesn't seem to connect directly to FCLK on this device, and routing is guaranteed to succeed.
-                continue;
-            }
-
-            int alias_count = 0;
-            std::set<std::set<BelId>> seen_options;
-            for (auto user : hclk_net->users) {
-                std::vector<BelId> bel_candidates;
-                std::set<BelId> these_options;
-
-                if (!(user.port == id_FCLK && (is_iologici(user.cell) || is_iologico(user.cell)) &&
-                      !user.cell->type.in(id_ODDR, id_ODDRC, id_IDDR, id_IDDRC)))
+        if (is_hclk_user(ci)) {
+            get_nets(ci);
+            for (auto &hclk_net : net_list) {
+                if (!hclk_net) {
                     continue;
-                if (seen_hclk_users.find(user.cell->name) != seen_hclk_users.end())
+                }
+                CellInfo *hclk_driver = hclk_net->driver.cell;
+                if (!hclk_driver)
+                    continue;
+                if (!(chip_has_clkdiv_hclk_connection || hclk_driver->type == id_CLKDIV2)) {
                     continue;
-                seen_hclk_users.insert(user.cell->name);
-
-                if (ctx->debug) {
-                    log_info("Custom HCLK Placer: Found HCLK user: %s\n", user.cell->name.c_str(ctx));
                 }
 
-                gwu.find_connected_bels(user.cell, id_FCLK, id_CLKDIV2, id_CLKOUT, 16, bel_candidates);
-                these_options.insert(bel_candidates.begin(), bel_candidates.end());
+                int alias_count = 0;
+                std::set<std::set<BelId>> seen_options;
```

Comment:
Can you explain exactly what this set-of-a-set is doing? I'm not sure I fully understand

Reply:
The internal set is CLKDIV2 Bels, which could potentially be used to deliver the signal to sinks located in the service area of the HCLK sections. For example PLL inputs can use sections 1 and 3, and IOLOGIC FCLK inputs from the same network sections 0 and 1, but individual IOs only section 1. 

The external set, on the other hand, simply maintains the uniqueness of these variant Bels sets.

My changes to the existing algorithm were only to ensure proper handling of the case where a single cell is a sink in multiple networks (pins connected to different HCLK networks). I can ask the author of the original commit of this algorithm to comment if there are more questions about the routing method itself.